types/testingutils/compare.go

```diff
@@ -96,33 +94,50 @@ func ComparePartialSignatureMessages(msg1 *types.PartialSignatureMessages, msg2
 		return errors.New("different messages length")
 	}
 
-	// messages are not guaranteed to be in order so we map them and then test all roots to be equal
-	roots := make(map[string]string)
-	for i, partialSigMsg2 := range msg2.Messages {
-		r2, err := partialSigMsg2.GetRoot()
+	// messages are not guaranteed to be in order so we map their roots and then test all roots to match and have the same multiplicity
+	roots1 := make(map[string]int)
+	roots2 := make(map[string]int)
+
+	// Count roots from msg1
+	for _, partialSigMessage := range msg1.Messages {
+		r1, err := partialSigMessage.GetRoot()
 		if err != nil {
 			return err
 		}
-		if _, found := roots[hex.EncodeToString(r2[:])]; !found {
-			roots[hex.EncodeToString(r2[:])] = ""
-		} else {
-			roots[hex.EncodeToString(r2[:])] = hex.EncodeToString(r2[:])
+		root := hex.EncodeToString(r1[:])
+		if _, found := roots1[root]; !found {
+			roots1[root] = 0
 		}
+		roots1[root] += 1
+	}
 
-		partialSigMsg1 := msg1.Messages[i]
-		r1, err := partialSigMsg1.GetRoot()
+	// Count roots from msg2
+	for _, partialSigMessage := range msg2.Messages {
+		r2, err := partialSigMessage.GetRoot()
 		if err != nil {
 			return err
 		}
 
-		if _, found := roots[hex.EncodeToString(r1[:])]; !found {
-			roots[hex.EncodeToString(r1[:])] = ""
-		} else {
-			roots[hex.EncodeToString(r1[:])] = hex.EncodeToString(r1[:])
+		root := hex.EncodeToString(r2[:])
+		if _, found := roots2[root]; !found {
+			roots2[root] = 0
 		}
+		roots2[root] += 1
 	}
```

Comment:
there is some copy paste of code... you can use a function

Reply:
True. Thankss